src/backend/common/CommandBufferStateTracker.cpp

```diff
@@ -0,0 +1,455 @@
+        if (currentRenderPass != nullptr) {
+            builder->HandleError("A render pass is already active");
+            return false;
+        }
+        if (!framebuffer->GetRenderPass()->IsCompatibleWith(renderPass)) {
+            builder->HandleError("Framebuffer is incompatible with this render pass");
+            return false;
+        }
+
+        currentRenderPass = renderPass;
+        currentFramebuffer = framebuffer;
+        currentSubpass = 0;
+        if (!BeginSubpass()) {
+            return false;
+        }
+
+        aspects.reset(VALIDATION_ASPECT_COMPUTE_PIPELINE);
+        aspects.reset(VALIDATION_ASPECT_RENDER_PIPELINE);
+        return true;
+    }
+
+    bool CommandBufferStateTracker::AdvanceSubpass() {
+        if (currentRenderPass == nullptr) {
+            builder->HandleError("Can't advance subpass without an active render pass");
+            return false;
+        }
+        if (currentSubpass + 1 >= currentRenderPass->GetSubpassCount()) {
+            builder->HandleError("Can't advance beyond the last subpass");
+            return false;
+        }
+
+        EndSubpass();
+        currentSubpass += 1;
+        if (!BeginSubpass()) {
+            return false;
+        }
+
+        aspects.reset(VALIDATION_ASPECT_RENDER_PIPELINE);
+        return true;
+    }
+
+    bool CommandBufferStateTracker::EndRenderPass() {
+        if (currentRenderPass == nullptr) {
+            builder->HandleError("No render pass is currently active");
+            return false;
+        }
+        if (currentSubpass < currentRenderPass->GetSubpassCount() - 1) {
+            builder->HandleError("Can't end a render pass before the last subpass");
+            return false;
+        }
+        EndSubpass();
+        currentRenderPass = nullptr;
+        currentFramebuffer = nullptr;
+
+        aspects.reset(VALIDATION_ASPECT_RENDER_PIPELINE);
+        return true;
+    }
+
+    bool CommandBufferStateTracker::SetPipeline(PipelineBase* pipeline) {
+        PipelineLayoutBase* layout = pipeline->GetLayout();
+
+        if (pipeline->IsCompute()) {
+            if (currentRenderPass) {
+                builder->HandleError("Can't use a compute pipeline while a render pass is active");
+                return false;
+            }
+            aspects.set(VALIDATION_ASPECT_COMPUTE_PIPELINE);
+        } else {
+            if (!currentRenderPass) {
+                builder->HandleError("A render pass must be active when a render pipeline is set");
+                return false;
+            }
+            if (!pipeline->GetRenderPass()->IsCompatibleWith(currentRenderPass)) {
+                builder->HandleError("Pipeline is incompatible with this render pass");
+                return false;
+            }
+            aspects.set(VALIDATION_ASPECT_RENDER_PIPELINE);
+        }
+        aspects.reset(VALIDATION_ASPECT_BIND_GROUPS);
+        aspects.reset(VALIDATION_ASPECT_VERTEX_BUFFERS);
```

Comment:
Previous code was doing it, however we shouldn't reset vertex buffers here: all backing APIs keep the vertex buffers bindings on pipeline change.